shuyangzhou/liferay-portal
# [Sidecar] LPS-103577

Hi Shuyang,

This pulls includes all the features required by sidecar.

Steps to test:

1. Run ant -f build-test-elasticsearch7.xml start-elasticsearch to build ES manually, will see a folder created in tomcat bundle home folder.
2. Rename the folder to elasticsearch7, and remove osgi/configs/com.liferay.portal.search.elasticsearch7.configuration.ElasticsearchConfiguration.config.
3. Start portal, login and go to control panel->configuration->search, will see `Search Engine Vendor: Elasticsearch (Embedded), Client Version: 7.3.0, Nodes: LiferayElasticsearchCluster: [liferay (7.4.1)]`

To test cluster,
Just enable cluster by setting "cluster.link.enabled=true"

Known issues:

1. Random failures on ServletContextHelperRegistrationImpl when shutting down, not sure if this is new or just triggered by my change, see https://gist.github.com/tinatian/8ae8776438a065be7fd2c677418ea9c6.
2. Elasticsearch does not allow to run on root user, needs to bypass
3. Error logging about log4j when starting 
`ERROR StatusLogger No Log4j 2 configuration file found. Using default configuration (logging only errors to the console), or user programmatically provided configurations. Set system property 'log4j2.debug' to show Log4j 2 internal initialization logging. See https://logging.apache.org/log4j/2.x/manual/configuration.html for instructions on how to configure Log4j 2
ERROR StatusLogger No Log4j 2 configuration file found. Using default configuration (logging only errors to the console), or user programmatically provided configurations. Set system property 'log4j2.debug' to show Log4j 2 internal initialization logging. See https://logging.apache.org/log4j/2.x/manual/configuration.html for instructions on how to configure Log4j 2
`

4. Clean up the configuration file, remove unused properties
5. Add unit/integration test
6. Test on different operation system

cc @arboliveira, I moved the impl of embedded connection to test cause a lot of unit tests need it. I am not sure if the test needs all the features, just simplify as possible as I can. Also, I kept most of the configs from embedded mode except those causing failures or not needed.

Thanks.
Tina.
